pkg/block/indexheader/parallel_bucket.go

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package indexheader
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+	"github.com/thanos-io/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"golang.org/x/sync/errgroup"
+)
+
+// partitionSize is used for splitting range reads for index-header.
+const partitionSize = 16 * 1024 * 1024 // 16 MiB
+
+type parallelBucketReader struct {
+	bkt           objstore.BucketReader
+	tmpDir        string
+	partitionSize int64
+}
+
+func WrapWithParallel(b objstore.BucketReader, tmpDir string) objstore.BucketReader {
+	return &parallelBucketReader{
+		bkt:           b,
+		tmpDir:        tmpDir,
+		partitionSize: partitionSize,
+	}
+}
+
+// GetRange implements ParallelBucket.
+func (b *parallelBucketReader) GetRange(ctx context.Context, name string, off int64, length int64) (io.ReadCloser, error) {
+	// id
+	partFilePrefix := uuid.New().String()
+	g := errgroup.Group{}
+
+	numParts := length / b.partitionSize
+	if length%b.partitionSize > 0 {
+		// A partial partition is remaining
+		numParts += 1
+	}
+
+	parts := make([]Part, 0, numParts)
+
+	i := 0
+	for o := off; o < off+length; o += b.partitionSize {
+		l := b.partitionSize
+		if o+l > off+length {
+			l = length - (int64(i) * b.partitionSize)
+		}
+
+		partOff := o
+		partLength := l
+		partId := i
+		part, err := b.createPart(partFilePrefix, partId, int(partLength))
+		if err != nil {
+			return nil, err
+		}
+		parts = append(parts, part)
+
+		g.Go(func() error {
+			rc, err := b.bkt.GetRange(ctx, name, partOff, partLength)
+			defer runutil.CloseWithErrCapture(&err, rc, "close object")
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("getRangePartitioned %v", partId))
+			}
+			buf := make([]byte, 32*1024)
+			if _, err := io.CopyBuffer(part, rc, buf); err != nil {
```

Comment:
I believe that we can use `io.Copy` because `io.File` implements the `ReaderFrom` interface which on linux systems will use a syscall to do the read:
* https://github.com/golang/go/blob/master/src/os/file.go#L149
* https://github.com/golang/go/blob/master/src/os/readfrom_linux.go#L18